src/System.Net.Requests/src/System/Net/FtpWebRequest.cs

```diff
@@ -0,0 +1,2544 @@
+                                {
+                                    m_RemainingTimeout = Timeout - (int)((DateTime.UtcNow - m_StartTime).TotalMilliseconds);
+                                    if(m_RemainingTimeout <= 0){
+                                        throw;
+                                    }
+                                }
+                            }
+                            continue;
+                        }
+                        throw;
+                    }
+                    // no retry needed
+                    break;
+                }
+            } catch (WebException webException) {
+                //if this was a timeout, throw a timeout exception
+                IOException ioEx = webException.InnerException as IOException;
+                if(ioEx != null){
+                    SocketException sEx = ioEx.InnerException as SocketException;
+                     if(sEx != null){
+                        if (sEx.SocketErrorCode == SocketError.TimedOut) {
+                            SetException(new WebException(SR.net_timeout, WebExceptionStatus.Timeout));
+                        }
+                    }
+                }
+
+                SetException(webException);
+            }
+            catch (Exception exception) {
+                SetException(exception);
+            }
+        }
+
+        private Exception TranslateConnectException(Exception e)
+        {
+            SocketException se = e as SocketException;
+            if (se != null)
+            {
+                if (se.SocketErrorCode == SocketError.HostNotFound)
+                {
+                    return new WebException(SR.net_webstatus_NameResolutionFailure, WebExceptionStatus.NameResolutionFailure);
+                }
+                else
+                {
+                    return new WebException(SR.net_webstatus_ConnectFailure, WebExceptionStatus.ConnectFailure);
+                }
+            }
+
+            // Wasn't a socket error, so leave as is
+            return e;
+        }
+
+        private FtpControlStream CreateConnection()
+        {
+            string hostname = m_Uri.Host;
+            int port = m_Uri.Port;
+
+            TcpClient client = new TcpClient();
+
+            try
+            {
+                if (m_Async)
+                {
+                    Task t = client.ConnectAsync(hostname, port);
+                    if (!t.IsCompleted)
+                    {
+                        t.ContinueWith((t2) =>
+                        {
+                            if (t2.IsFaulted)
+                            {
+                                AsyncRequestCallback(TranslateConnectException(t2.Exception));
+                            }
+                            else
+                            {
+                                AsyncRequestCallback(new FtpControlStream(client));
+                            }
+                        });
```

Comment:
> That said, it seems to me that async/await is going to post back to the current scheduler/synccontext, right? Is that a problem?

Yes, it can deadlock. It's easily fixed by appending a `.configureawait(false)` to all internal `await` statements.